ENH add button to GUI for allowing EEG as head points

cjayb opened the pull request:
Thanks for the great GUI @christianbrodbeck 
I wonder if there's a reason not to include the option of using digitised EEG locations as head points during coregistration? I understand that this is conditional on the digitisation pen being stuck through the electrode ring all the way to the scalp, but would it make sense to allow the user to tick a box to use them?

I've made the few edits needed to expose the option in the GUI, but am having trouble understanding Traits. As far as I can tell, the points returned by _get_raw_points and the new _get_eeg_points could be concatenated and the rest of the code remain as-is. If you agree the feature might be useful, I'd appreciate some pointers on how to proceed, and perhaps a word on testing.

christianbrodbeck commented on mne/gui/_coreg_gui.py:
```diff
@@ -1348,6 +1352,10 @@ def _get_hsp_visible(self):
 
     @cached_property
+    def _get_eeg_as_hsp(self):
+        return self.use_eeg_as_hsp
```
This is not necessary. These `_get_x` methods are for properties that are defined with `depends_on`, i.e. if you define a property that depends on others you use these functions to tell Traits how they're generated.

christianbrodbeck commented:
It looks like the HSP points are provided by the `hsp` trait, which is a `RawSource` instance. The points come from `hsp.points`.

Probably first you'd want to add a new trait to `RawSource` that exposes the EEG locations, e.g. `RawSource.eeg_points`. Then in the CoregModel create an intermediary `points` trait that depends_on the HSP points, the EEG points and the setting on whether to include the EEG points. Then you can just replace references to `hsp.points` with references to the intermediary points object. Let me know if you need more help.

agramfort commented:
what's our status here?

cjayb commented:
Status is stalled, I'm afraid. I won't have proper time for this until I return to work in early Aug. I think I understand the basic idea Christian suggests, and can easily get the GUI elements in. It's what to do with the EEG points once they're added to the list of head points that needs more thought. I think this is an important feature, though, since it's hard to get a good set of "additional points" using the Polhemus once the cap is on. If no-one else picks this up, I'll surely work on it as soon as I can.

cjayb commented:
Hi, just a brief status update (no need to break out of holiday-mode for this...)

I've rebased and changed my approach. I've decided to use the `points_filter` mechanism for omitting EEG points, and the last commit (@b4a9a55) indeed indicates that this works using two new buttons. However, I think I would like to try to work on a slightly different approach (than cluttering buttons): to augment the definition of an omission filter. 

When using an EEG cap, getting "true" head shape points from the scalp is tricky, and thus I recommend users of our 75 ch EEG caps that they simply don't bother. It's customary to mark points on the face (bridge of nose, eyebrows) as sanity checks that the coreg is working, but I would prefer these not to bias the automatic fit. So, in fact, an "omit extra points" would also be nice. Finally, sometimes the MR images have really poor facial contrast (either due to choice of head coil or to "anonymisation"), and one might want to omit based on, say, the y-coordinate.

All these features can potentially be implemented in some sort of a filter-dialogue, which I'll try to take a look at and present a suggestion. Comments are welcome.

agramfort commented:
whatever works but don't overfit too much your usecase. You may also add a
param in the script or function call rather than adding buttons if it's
always the same action.

cjayb commented:
I'm afraid I've crawled too deep into this rabbit hole and need help getting out again! I decided to go for the "simple" approach of in/excluding EEG points using a tick box. I found that in order for this to work with `omit_hsp_points` I had to modify the latter slightly by adding a call to the new `_apply_eeg_filter` method (L343).

I'd like `use_eeg_locations` to be a parameter of the `model`, so I define it there (L159) and `DelegateTo` it in the `CoregFrame` (L1217). This seems to work logically, as the TraitsUI element generated is indeed a check box. Using various `print`-statements for debugging, I'm convinced that the selection actually works (though see next paragraph).

`_get_points` on L359 of `_file_traits.py`: I had to remove the `@cached_property`-decorator. If I don't, it seems that accessing `hsp.points` does not obey the dependency on `hsp.points_filter`, even though this is defined on L298. What could have caused this to break?!

Finally, my most serious bug is that the scene showing the head points does not update to the toggling of `use_eeg_locations`. So even though debugging indicates that (with the above caveat in `_get_points`) the `hsp.points_filter` is updated correctly, the point cloud does not reflect this.

I'll try to find the time to debug this myself, but have already spent a considerable amount of time and wonder if Traits are in fact healthy for one's mental well-being...?

Eric89GXL commented:
I find `traits` to be pretty difficult to work with, so I can't really help. Maybe @christianbrodbeck would have some ideas?

christianbrodbeck commented:
Yeah and I've found them hard to debug because so much happens behind the scenes...

For the first issue, could it be due to a non linear dependency? Now inst_points and points_type both depend on inst, and points depends on those two. I remember that that kind of nonlinear dependence could be problematic, possibly a trait can't be invalidated twice in the same cycle? I would try rewriting depends_on such that one change triggers the next.

For the second I noticed you are assigning an array on L351. In order for a cached_property to automatically register a change in a dependency, IIRC the object has to change, not just its content, i.e. if it is an array it is not enough to just change its values, the array object has to be replaced. Might this be related to the failure to update? There might be a way to trigger the change with an invalidate–property function or so.

cjayb commented:
Thanks all for the input. @christianbrodbeck I cleaned up the cross-dependencies, but it was object replacement that did the trick. I didn't find any mention of an `invalidate_property`-magic, so I forced an array copy both in the EEG-specific filter and the distance-based omission function. (The latter had the side-effect of fixing the strange behaviour in `master` when clicking on `Omit` twice...)

I'd like to submit this for review. I believe my bits passed through tests, but there's an unrelated conflict somewhere in `whats_new.rst`(?).

This has been an excellent nudge to focus on `ipynb` widgets for small-scale GUIs, and perhaps to learn some d3 or bokeh for bigger projects :)

christianbrodbeck commented:
@cjayb I'm traveling currently but will be able to review it around September 1 or 2 if it's still open then.

christianbrodbeck commented on mne/gui/_file_traits.py:
```diff
@@ -334,13 +338,24 @@ def _get_inst_fname(self):
     def _get_inst_points(self):
         if not self.inst:
-            return np.zeros((1, 3))
+            return np.zeros((1, 3))  # why not just None?
```
I don't remember, but setting it to None causes an exception on startup, so it would require downstream modification to allow it to be None. Simpler might be setting its shape to (0, 3), that seems to work for me. I don't know why (1, 3).

cjayb replied:
OK, changed it to (0,3) and seems to work

christianbrodbeck commented on mne/gui/_coreg_gui.py:
```diff
@@ -106,6 +106,7 @@ class CoregModel(HasPrivateTraits):
     has_fid_data = Property(Bool, depends_on=['mri_origin', 'hsp.nasion'],
                             desc="Required fiducials data is present.")
     has_pts_data = Property(Bool, depends_on=['mri.points', 'hsp.points'])
+    dig_points = Property(depends_on=['hsp.inst_points'])
```
I don't think you're using this trait below, right? Maybe it's left over from an earlier implementation?

cjayb replied:
yup, deleted

christianbrodbeck commented:
Otherwise the code looks good to me. Do you want to add a test? I just checked and the testing dataset does contain EEG points. 

A minor and optional point, It might be nice for the text right below the new check box to indicate the number of points excluded because they're EEG and the ones excluded based on distance separately, but I don't think that's necessary, only if you feel like it :).

cjayb commented:
Thanks @christianbrodbeck I've addressed your comments and included a possible solution on how to indicate which points have been excluded/omitted. Let me know if you agree...

Yes, I'd like to add a test, need a bit more time to consider exactly what to test...

cjayb commented:
I simply made partitioned testing to be with-then-without EEG electrodes as HSPs. 

I suppose we could test for things like `assert_that(points_of_type_EEG==info['channels of type EEG'])`. Think that's needed?

christianbrodbeck commented on mne/gui/_coreg_gui.py:
```diff
@@ -1352,16 +1382,26 @@ def _get_omitted_info(self):
+        self.model.apply_eeg_filter()  # Don't forget the EEG
+
+    def _use_eeg_locations_fired(self):
+        self.model.apply_eeg_filter()
```
Couldn't `_use_eeg_locations_fired` be on the `CoregModel`? Then you would not have to manually trigger it in the tests?

christianbrodbeck commented:
@cjayb yes I think it would be good to test that the number of points changes, that should be more reliable than checking that the fit improves, I think you can just hard code the desired number of channels.

cjayb commented:
Yes, `points_type` should indeed be a 1D-array. Good idea refactoring the `_use_eeg_locations_fired` to the model (and moving it to `_use_eeg_locations_changed`).

I've made some tests to check the various behaviours and interactions of Omit, Reset and the EEG checkbox. IMHO it's working now. Happy with the tests?

christianbrodbeck commented:
The updates look good, I just noticed one unexpected behavior though: 

- set the MRI
- load Head Shape Source
- uncheck "use EEG electrode ..."
- load a different Head Shape Source

->  "use EEG electrode ..." is now unchecked but no points are excluded

This might become relevant when using average brains, because then you use the same MRI for coregistering multiple data files... 

I'm having some trouble thinking of what trait change to listen to to fix this... maybe the simplest solution would be to move the whole EEG include/exclude logic into the `InstSource` and apply it inside `InstSource._get_points()`? That should cause the points that are used by the `CoregModel` to automatically always reflect the parameter. (i.e. also add the `.use_eeg_locations` trait into `InstSource`, and the `CoregModel` can `delegate_to` it)

cjayb commented:
Bugger, I see the problem. You may be right that the only way to get a trigger is to include the Boolean flag in the hsp-Instance, but let me sleep on it. I dislike the idea of sticking it there as it's really not a property of the data source but of the model. maybe there's another way...

christianbrodbeck commented:
> I dislike the idea of sticking it there as it's really not a property of the data source but of the model

I had the same intuition, and was wondering whether it might make more sense to put the whole point exclusion (including `points_filter`) into the the `CoregModel` instead... I just checked and `InstSource` is not used anywhere else, so it should even be OK to change it and remove those elements...

cjayb commented:
Makes sense to me. I hope to get some time over the weekend to make a proposal.

cjayb commented:
Well, I made it, I think... The most prominent hacks are in the line notes, let me know what you think.

christianbrodbeck commented on mne/gui/_coreg_gui.py:
```diff
@@ -153,12 +167,57 @@ class CoregModel(HasPrivateTraits):
+        else:
+            return self.hsp.inst_points[self.points_filter]
+
+    def _get_points_filter(self):
```
I think this method was left here by accident?

christianbrodbeck replied:
This method (`._get_points_filter()`) is identical to `_get_points()`, and is never called because `points_filter` is not a Property.

cjayb replied:
DUH! Gimme a second...

christianbrodbeck commented on mne/gui/_coreg_gui.py:
```diff
@@ -80,6 +80,20 @@ class CoregModel(HasPrivateTraits):
+    n_omitted = Property(Int, depends_on=['points_filter'])
+    n_omitted_types = Property(depends_on=['n_omitted'])
+
+    points = Property(depends_on=['inst_points', 'points_filter'],
```
why not add `use_eeg_locations` to `depends_on` and apply it in `._get_points()`?

christianbrodbeck commented on mne/gui/_coreg_gui.py:
```diff
@@ -80,6 +80,20 @@ class CoregModel(HasPrivateTraits):
+    points_filter = Any(desc="Index to select a subset of the head shape "
+                        "points")
+    n_omitted = Property(Int, depends_on=['points_filter'])
+    n_omitted_types = Property(depends_on=['n_omitted'])
```
I think these could be combined into a single trait, would be simpler

christianbrodbeck commented:
Yeah if there isn't something I missed I think `use_eeg_locations` should be applied in `_get_points`, that would be more "traity"... if you're tired of this by now you can let me know and I can make a PR into your PR :)

cjayb commented:
No way I'm going to let you steal my thunder ;) Seriously, it's a great learning experience for me. I combined the various omitted info's into one struct. This made counting the total number of `n_omitted` a little peculiar (`sum(....values())`), but I think it's better this way. Will also be easy to extend if you _e.g._ want to add HPI as a type of head point (apropos recent mailing list activity).

I didn't get your point about `_get_points`, though. I think it should be where it is (`points` is now a trait of the model). Maybe you meant `_get_points_filter` should `depend_on` `use_eeg_locations`? I'll take a look, but I don't have anything against a PR on my PR, the more the merrier :)

cjayb commented:
Right, so I finally saw (and removed) the duplication, but still cannot quite fathom how to clear up some of the EEG logic?

christianbrodbeck commented:
What I meant with the other comment was that with Traits, you try to keep the logic in `Property` attributes with associated  `_get_x()` methods, and avoid methods that are called from different places. Wouldn't it be possible in this case to avoid the `apply_eeg_filter()` method by moving its logic into `._get_points()`? If you add `"use_eeg_locations"` to `points`' `depends_on`, the points should then always properly reflect it.

cjayb commented:
Ah, yes, perhaps... I think I managed to _almost_ get it to work in this branch/commit: a5c67b33bca5dd93e0c69875aa76afe5895e62c2

The problem is that since `omitted_info` `depends_on` `points_filter`, I end up with an infinite recursion if I force-invalidate the filter in the `get_points`-method. But if I don't the info string doesn't update in the GUI. Any ideas?

I'm not sure how to git-deal with this situation: should I merge the commit even though it might mean a rollback later? Can you see the commit to test it? I'll be back tomorrow...

christianbrodbeck commented:
> I'm not sure how to git-deal with this situation...

Yeah you can always change the history of a PR, just use `git push -f` and the PR reflects your new history.

> infinite recursion

hm right, you need a dependence structure without loops... since `n_used` updates always together with `points` you could just directly create and assign the n_omitted (or even the omitted info string) in `_get_points()`, no?

Then `points` would depend on `points_filter` (determined only by user action with the omit-points function) and the `use_eeg_pooints` setting and would also not have to re-assign `points_filter`. 

Does that make sense?

cjayb commented:
You are quite right, Sir. I very much like the way `_get_points` is the central hub for updating the traits that reflect the current set of omissions (filer and info-dict). I was a little afraid about making the method overly complicated, but I think it ended up being just the opposite!

christianbrodbeck commented on mne/gui/_coreg_gui.py:
```diff
@@ -80,6 +80,25 @@ class CoregModel(HasPrivateTraits):
+    # head shape points relevant to the Model depend on the filter
+    points_filter = Any(desc="Index to select a subset of the head shape "
+                        "points")
+    omitted_info = Any(desc="Dictionary with number of various points omitted")
```
Should probably make this a `DictStrInt` (http://code.enthought.com/projects/files/ETS3_API/enthought.traits.trait_types.html) and assign a default value including the keys read by the CoregFrame (`DictStrInt(..., value={'n_total': 0, ...})`, see http://code.enthought.com/projects/files/ETS3_API/enthought.traits.trait_types.Dict.html). Now I get this when launching the GUI (I think for Any the value at initialization is None): 

```
  File "/Users/christian/Code/mne-python/mne/gui/_coreg_gui.py", line 1407, in _get_omitted_string
    if self.model.omitted_info['n_total'] == 0:
TypeError: 'NoneType' object has no attribute '__getitem__'
```

cjayb commented:
Yes, much cleaner with the `DictStrInt`, now with sane default `value`. For the record, I didn't get the `TypeError` you did when starting the GUI (wouldn't have pushed if I had). Scary...

I'm seeing 4 uncaught warnings during the nosetests of coreg_gui:
`FutureWarning: comparison to `None` will result in an elementwise object comparison in the future.`
but could figure out where they were coming from to squash them.

Eric89GXL commented:
Those are probably coming from Traits

Eric89GXL commented:
PR sent for warnings:

https://github.com/cjayb/mne-python/pull/2

cjayb commented:
Ha! I had figured that was the function of `catch_warnings` but forgot to `with` them in the test-script :) Thanks @Eric89GXL

christianbrodbeck commented:
Hm, now when I try to start it I get 

```
$ mne coreg
Traceback (most recent call last):
  File "/Users/christian/Code/mne-python/mne/commands/mne_coreg.py", line 29, in <module>
    run()
  File "/Users/christian/Code/mne-python/mne/commands/mne_coreg.py", line 23, in run
    mne.gui.coregistration()
  File "/Users/christian/Code/mne-python/mne/gui/__init__.py", line 65, in coregistration
    gui.configure_traits(view=view)
  File "//anaconda/lib/python2.7/site-packages/traits/has_traits.py", line 2386, in configure_traits
    kind, handler, id, scrollable, args )
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/toolkit.py", line 212, in view_application
    id, scrollable, args )
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/view_application.py", line 83, in view_application
    scrollable, args ).ui.result
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/view_application.py", line 134, in __init__
    args       = self.args )
  File "//anaconda/lib/python2.7/site-packages/traitsui/view.py", line 433, in ui
    ui.ui( parent, kind )
  File "//anaconda/lib/python2.7/site-packages/traitsui/ui.py", line 221, in ui
    self.rebuild( self, parent )
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/toolkit.py", line 154, in ui_live
    ui_live.ui_live( ui, parent )
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_live.py", line 39, in ui_live
    _ui_dialog(ui, parent, BaseDialog.NONMODAL)
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_live.py", line 59, in _ui_dialog
    BaseDialog.display_ui(ui, parent, style)
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_base.py", line 445, in display_ui
    ui.owner.init(ui, parent, style)
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_live.py", line 177, in init
    self.add_contents(panel(ui), bbox)
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_panel.py", line 231, in panel
    panel = _GroupPanel(content[0], ui).control
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_panel.py", line 524, in __init__
    self._add_splitter_items(content, splitter)
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_panel.py", line 603, in _add_splitter_items
    panel = _GroupPanel(item, self.ui, suppress_label=True).control
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_panel.py", line 570, in __init__
    layout = self._add_groups(content, inner)
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_panel.py", line 649, in _add_groups
    panel = _GroupPanel(subgroup, self.ui).control
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_panel.py", line 570, in __init__
    layout = self._add_groups(content, inner)
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_panel.py", line 649, in _add_groups
    panel = _GroupPanel(subgroup, self.ui).control
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_panel.py", line 572, in __init__
    layout = self._add_items(content, inner)
  File "//anaconda/lib/python2.7/site-packages/traitsui/qt4/ui_panel.py", line 832, in _add_items
    ui, object, name, item.tooltip, None
  File "//anaconda/lib/python2.7/site-packages/traitsui/editor_factory.py", line 170, in readonly_editor
    description = description )
  File "//anaconda/lib/python2.7/site-packages/traitsui/editor.py", line 133, in __init__
    self.old_value = getattr( self.object, self.name )
  File "//anaconda/lib/python2.7/site-packages/traits/has_traits.py", line 1126, in decorator
    self.__dict__[ name ] = result = function( self )
  File "/Users/christian/Code/mne-python/mne/gui/_coreg_gui.py", line 1413, in _get_omitted_string
    if self.model.omitted_info['n_total'] == 0:
KeyError: 'n_total'
```

It looks like omitted info is initialized as `{}` and only then set to the default... if I insert the following into `_get_omitted_str`:

```
        print self.model.omitted_info
        if not self.model.omitted_info:
            return ""
```

I get 

```
$ mne coreg
{}
{'n_total': 0, 'n_eeg': 0, 'n_hsp': 0}
```

Strange that you don't get that, I am on OS X with anaconda and

```
$ pip show traits
---
Metadata-Version: 1.1
Name: traits
Version: 4.4.0
```

Other than that the PR works great now. I noticed that "Omit Points" does not omit any EEG points, but that might be the desired behavior since EEG points would be wrong if they were not on the head, right?

cjayb commented:
Very odd, I get no `KeyError` on startup (obviously) and with your sanity check print statements, merely:

```
{'n_total': 0, 'n_eeg': 0, 'n_hsp': 0}
```

(_i.e._, no `{}`). And all this on OS X with a conda install of traits and traitsui, both at

```
$ pip show traits
---
Metadata-Version: 1.1
Name: traits
Version: 4.4.0
```

I'm at a loss here...

cjayb commented:
_I noticed that "Omit Points" does not omit any EEG points, but that might be the desired behavior since EEG points would be wrong if they were not on the head, right?_

Actually, this is a bug as far as I'm concerned :( It's highly likely that not all electrodes are optimally digitised on the scalp, but are left "floating". In these cases omitting by distance is just as desirable as with "normal" HSPs. 

I've struggled to see how to implement a better behaviour, but am a bit stumped. I'll have to take a fresh look  a little later. The problem is L191 in `_coreg_gui.py`

```self.points_filter[eeg_filter] = self.use_eeg_locations```

which effectively overrides any changes to the `points_filter` coming from the omission logic. I thought I had an easy fix, but I can't get past the need for an auxiliary variable doing some bookkeeping on the "history" of the filter. Ends up being too untidy code to commit...

christianbrodbeck commented:
> KeyError

Really strange, I get the same error on my laptop too.

> Actually, this is a bug as far as I'm concerned :( It's highly likely that not all electrodes are optimally digitised on the scalp, but are left "floating". 

In this case wouldn't you actually have to go back and fix the electrode position because otherwise you could not use it in source localization?

christianbrodbeck commented on mne/gui/_coreg_gui.py:
```diff
@@ -159,6 +180,38 @@ class CoregModel(HasPrivateTraits):
+        with warnings.catch_warnings(record=True):  # comp to None in Traits
+            if self.points_filter is None:
+                self.points_filter = \
+                    np.ones(len(self.hsp.inst_points), np.bool8)
```
The only `Property` that depends on `points_filter` is `points`, so `_get_points()` should not reassign `points_filter`, it should just read it.

Couldn't you just creat an index by
 - starting with `points_filter` (or if it's `None` uses `np.ones()`), 
 - and if EEG electrodes should be excluded use `np.logical_and`

or am I missing something?

cjayb replied:
I thought of that too (`np.logical_and`), but consider this:
* load a hsp source, EEG enabled by default
* disable EEG
* try to Omit by distance

Now when `omit_hsp_points` gets `self.transformed_hsp_points`, `_get_points` returns _fewer points_ (because the EEG points are missing) than are defined in the `points_filter`. Then this line fails due to the mismatch in tested points (`new_sub_filter`) vs. those defined to be True in the `old_filter`:

```
new_filter[old_filter] = new_sub_filter
```

Making `omit_hsp_points` aware of the EEG selector state seems clumsy and prone to future bugs. If I do modify the filter in `_get_points`, though, I end up not being able to turn the EEG back on after they've been left out once (the logical and will forever be False).

christianbrodbeck replied:
Yeah hmmm I agree it's hard to find a solution that feels clean...  making `omit_hp_points` aware of the EEG selector state might be the best option (even though not perfect). To make it more transparent, maybe `_get_points()` could also assign a `Property` that is an index to project `points` back to the raw points. For example, with 3 points of the kind `[HSP, EEG, HSP]` the back-projector could be `[0, 2]`.

cjayb commented:
> In this case wouldn't you actually have to go back and fix the electrode position because otherwise you could not use it in source localization?

Umm, yes, good point. Simply haven't got to worrying about that yet... The electrodes need to be projected down to the scalp surface anyway at a later point (visualisation and/or forward modeling). Assuming that the omitted "floating" electrodes are mostly off in the radial direction (usually a pretty safe bet, I think), things should be reasonably OK.

Eric89GXL commented:
@cjayb @christianbrodbeck do either of you have time to come back to this? It has been open over a year so it would be good to get it in if it's useful

christianbrodbeck commented:
It seems to be a marginal use case, and turned out to be more complex than first anticipated. Unless there are multiple people that would benefit from this feature it might be more efficient to just manipulate the underlying FIFF files (duplicate the EEG points as digitizer points in the info dict)?

cjayb commented:
I managed to get by with the almost functional version I wrote, but never quite figured out how to iron out the few remaining rough edges. I'm rather surprised that this turns out to be a marginal use case, but given the obvious lack of requests from others, I guess that's the way it is.

The main problems arose in keeping the two different point types separate. Perhaps a new stab at a PR should take the route suggested by @christianbrodbeck i.e. hacking the info dict? Not as elegant, but then again mine wasn't either! I'll let you guys decide whether to close this, I don't have the resources right now to get to it, sorry.

Eric89GXL commented:
I'll close for now, feel free to resurrect it if anyone finds time and motivation to work on it
